File: ios/WorkoutRecord/WorkoutRecord/Sources/Models/PersonalRecord.swift
import Foundation

// MARK: - Personal Record

struct PersonalRecord: Identifiable, Codable {
    let id: UUID
    let userId: UUID
    let exerciseId: UUID
    var exercise: Exercise?
    var weight: Double        // PR 重量
    var reps: Int             // 次數
    var oneRepMax: Double     // 估算的 1RM
    let achievedAt: Date      // 達成日期
    let workoutId: UUID?      // 關聯的訓練 ID
    let createdAt: Date
    var updatedAt: Date
    
    init(
        id: UUID = UUID(),
        userId: UUID,
        exerciseId: UUID,
        exercise: Exercise? = nil,
        weight: Double,
        reps: Int,
        oneRepMax: Double,
        achievedAt: Date = Date(),
        workoutId: UUID? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.userId = userId
        self.exerciseId = exerciseId
        self.exercise = exercise
        self.weight = weight
        self.reps = reps
        self.oneRepMax = oneRepMax
        self.achievedAt = achievedAt
        self.workoutId = workoutId
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

// MARK: - PR Summary (按動作分組)

struct PRSummary: Identifiable {
    let id: UUID
    let exerciseId: UUID
    let exerciseName: String
    let primaryMuscleGroup: Exercise.PrimaryMuscleGroup?
    var currentPR: PersonalRecord?
    var prHistory: [PersonalRecord] = []
    
    var weight: Double? {
        currentPR?.weight
    }
    
    var reps: Int? {
        currentPR?.reps
    }
    
    var oneRepMax: Double? {
        currentPR?.oneRepMax
    }
    
    var achievedAt: Date? {
        currentPR?.achievedAt
    }
    
    init(
        id: UUID = UUID(),
        exerciseId: UUID,
        exerciseName: String,
        primaryMuscleGroup: Exercise.PrimaryMuscleGroup? = nil,
        currentPR: PersonalRecord? = nil,
        prHistory: [PersonalRecord] = []
    ) {
        self.id = id
        self.exerciseId = exerciseId
        self.exerciseName = exerciseName
        self.primaryMuscleGroup = primaryMuscleGroup
        self.currentPR = currentPR
        self.prHistory = prHistory
    }
}

File: app/lib/data/models/personal_record.dart
import 'package:drift/drift.dart' show Value;

import '../db/app_database.dart' as db;
import 'exercise.dart';

/// 個人紀錄(PR),對應 Drift `PersonalRecords` 表。
///
/// [exercise] 為選填的關聯動作資料(由 repository join `Exercises` 表填入)。
class PersonalRecord {
  final String id;
  final String userId;
  final String exerciseId;
  final Exercise? exercise;
  final double weight;
  final int reps;
  final double oneRepMax;
  final DateTime achievedAt;
  final String? workoutId;
  final DateTime createdAt;
  final DateTime updatedAt;

  const PersonalRecord({
    required this.id,
    required this.userId,
    required this.exerciseId,
    this.exercise,
    required this.weight,
    required this.reps,
    required this.oneRepMax,
    required this.achievedAt,
    this.workoutId,
    required this.createdAt,
    required this.updatedAt,
  });

  factory PersonalRecord.fromRow(db.PersonalRecord row, {Exercise? exercise}) {
    return PersonalRecord(
      id: row.id,
      userId: row.userId,
      exerciseId: row.exerciseId,
      exercise: exercise,
      weight: row.weight,
      reps: row.reps,
      oneRepMax: row.oneRepMax,
      achievedAt: row.achievedAt,
      workoutId: row.workoutId,
      createdAt: row.createdAt,
      updatedAt: row.updatedAt,
    );
  }

  db.PersonalRecordsCompanion toCompanion() {
    return db.PersonalRecordsCompanion(
      id: Value(id),
      userId: Value(userId),
      exerciseId: Value(exerciseId),
      weight: Value(weight),
      reps: Value(reps),
      oneRepMax: Value(oneRepMax),
      achievedAt: Value(achievedAt),
      workoutId: Value(workoutId),
      createdAt: Value(createdAt),
      updatedAt: Value(updatedAt),
    );
  }

  PersonalRecord copyWith({
    double? weight,
    int? reps,
    double? oneRepMax,
    DateTime? achievedAt,
    String? workoutId,
    DateTime? updatedAt,
  }) {
    return PersonalRecord(
      id: id,
      userId: userId,
      exerciseId: exerciseId,
      exercise: exercise,
      weight: weight ?? this.weight,
      reps: reps ?? this.reps,
      oneRepMax: oneRepMax ?? this.oneRepMax,
      achievedAt: achievedAt ?? this.achievedAt,
      workoutId: workoutId ?? this.workoutId,
      createdAt: createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

/// 依動作分組的 PR 總結(非資料表,由 [PersonalRecord] 聚合而成)。
/// 對照 iOS `PersonalRecordRepository.getPRSummary`。
class PRSummary {
  final String exerciseId;
  final String exerciseName;
  final PrimaryMuscleGroup? primaryMuscleGroup;
  final PersonalRecord? currentPR;
  final List<PersonalRecord> prHistory;

  const PRSummary({
    required this.exerciseId,
    required this.exerciseName,
    this.primaryMuscleGroup,
    this.currentPR,
    this.prHistory = const [],
  });

  double? get weight => currentPR?.weight;
  int? get reps => currentPR?.reps;
  double? get oneRepMax => currentPR?.oneRepMax;
  DateTime? get achievedAt => currentPR?.achievedAt;
}
